### DISCUSSION TITLE
在onPostTransform方法中如何修改源码呢，请指教

### DISCUSSION THREAD

**Discussion by hellokai55** [didi/booster#400] [Q&A] [answered]
我的需求是这样的：

针对于每一个module会生成一个Mapping_xxx.java，里面会有一个返回map的方法；想在编译期给他们聚合起来，写了一个源文件Mapping.java类，提供了一个inject()的空方法，想在这个方法里把遍历所有的符合条件的Mapping_xxx.java的map加到一起，运行时其他人直接调Mapping.java类就好了。

如果写一个自定义的ClassTransformer，那么在transform做寻找Mapping_xxx.java和Mapping.java的聚合就会有顺序问题；所以想要先遍历一遍把所有的Mapping_xxx.java都找到，然后再修改Mapping.java中的inject方法做聚合，能在onPostTransform中再修改Mapping.java类吗，该怎么做呢？

**Comment by johnsonlee** [comment, marked as answer]
You can use [Booster Collector API](https://reference.johnsonlee.io/booster/com.didiglobal.booster.transform/-collector/index.html) to collect information before the transform pipeline is executed.

Please refer to [Multi-Round Transformer](https://booster.johnsonlee.io/en/developer/multi-round-transform.html) to get more details.

**Reply by johnsonlee** [reply]
You can define an empty `Mapping` class as the facade, and then collect the generated `Mapping_xxx` files from each module, then generate bytecode according to the collected information during the `Mapping` class transforming, that's what [ServiceLoaderTransform](https://github.com/didi/booster/blob/master/booster-transform-service-loader/src/main/kotlin/com/didiglobal/booster/transform/serviceloader/ServiceLoaderTransformer.kt) did.

**Reply by hellokai55** [reply]
大佬这个,Booster Collector API 怎么用呢，没太看懂这个[Multi-Round Transformer](https://booster.johnsonlee.io/en/developer/multi-round-transform.html)使用方式，例如每一个module我都会生成文件例如：A的module类是：com.test.map_ + hash值；B的module是com.text.map_ + hash值；然后我有一个类作为聚合叫：com.test.map.Mapping；

伪代码：
```
class MyCollector(mapClass: String, val list: List<String>) : Collector<String> {

   //相匹配到符合com/test/DPRouterMapping_这个开头的文件
    override fun accept(name: String): Boolean = name.startWith(mapClass)

    override fun collect(name: String) =  name.apply {
        //怎么能把匹配到的所有文件加到list中呢？从实际来看不太对，有很多重复的类添加
        list.add(name)
    }

}


@AutoService(ClassTransformer::class)
class DPPageSchemeTransformer : ClassTransformer {
    companion object {
        const val TAG = "DPPageSchemeTransformer"
        const val MAPPER_CLASS = "com/test/DPRouterMapping_"
        const val TARGET_CLASS = "com/test/runtime/PageSchemeMapping"

    }

   val serviceName = mutableListOf<String>()

    override fun onPreTransform(context: TransformContext) {
        //我应该在这注册collector，想将所有的符合条件的都放到serviceName中
        context.registerCollector(MyCollector(MAPPER_CLASS, serviceName))
    }

    override fun transform(context: TransformContext, klass: ClassNode): ClassNode {
        //我在这里怎么才能拿到所有MyColletor中匹配到的文件列表呢？

        if (klass.name == TARGET_CLASS) {//聚合处理
            targetPageSchemeKlass = klass
            //修改字节码，serviceName中保存的就是我所有配到的数组，然后我改下TARGET_CLASS中的方法将数组聚合下
            println("$TAG targetPageSchemeKlass, ${klass.name}")
        }
        return klass
    }

}

```

大佬能帮我看看咋解决吗？

**Reply by johnsonlee** [reply]
参考 Booster 提供的 `ClassNameSupervisor`:

## MappingSupervisor

```kotlin
class MappingSupervisor(
        val list: MutableList<String>
) : AbstractSupervisor<String>({ list.add(it) }) {

    private companion object {
        private val MAPPING = "map_[a-fA-F0-9]+\\.class".toRegex()
    }

    override fun accept(name: String): Boolean {
        return name.substringAfterLast("/") matches MAPPING
    }

    override fun collect(name: String, data: () -> ByteArray) {
        action(name.substringBeforeLast(".class").replace('/', '.'))
    }

}
```

## DPPageSchemeTransformer

```kotlin
@AutoService(ClassTransformer::class)
class DPPageSchemeTransformer : ClassTransformer {

    val serviceNames = mutableListOf<String>()

    override fun onPreTransform(context: TransformContext) {
        context.registerCollector(MappingSupervisor(serviceNames))
    }

    override fun transform(context: TransformContext, klass: ClassNode) = klass.also {
        if (klass.name == TARGET_CLASS) {
            serviceNames.forEach { serviceName ->
                // TODO: Generate your byte code ...
            }
        }
    }

}
```

**Reply by hellokai55** [reply]
刚才我正好也在看ServiceSupervisor的类的实现，知道怎么写了，多谢大佬代码指导！
